refactor: Replace global find_value function with UniValue::find_value method

maflcko opened the pull request:
The global function has issues:

* It causes gcc-13 warnings, see https://github.com/bitcoin/bitcoin/issues/26926
* There is no rationale for it being a global function, when it acts like a member function
* `performance-unnecessary-copy-initialization` clang-tidy isn't run on it

Fix all issues by making it a member function.

Commit:
Add UniValue::find_value method

Commit:
scripted-diff: Use UniValue::find_value method

-BEGIN VERIFY SCRIPT-
 sed --regexp-extended -i 's/find_value\(([^ ,]+), /\1.find_value(/g' $(git grep -l find_value)
-END VERIFY SCRIPT-

Commit:
Remove unused find_value global function

Commit:
Fix clang-tidy performance-unnecessary-copy-initialization warnings

fanquake commented:
Concept ACK

hebasto commented:
Concept ACK.

hebasto reviewed (approved):
ACK fa0d180f48d4e5253f58aced32768369f963d1c7, tested on Ubuntu 23.04 using GCC 13.0.1.

stickies-v reviewed (approved):
ACK fa0d180f48d4e5253f58aced32768369f963d1c7

I can't see any behaviour change, and as far as I can tell using references in fa0d180f48d4e5253f58aced32768369f963d1c7 doesn't introduce any (potential) lifetime issues.

hebasto commented:
> > Ubuntu 23.04 using GCC 13.0.1
> 
> I think it ships with 12.2? [packages.ubuntu.com/lunar/gcc](https://packages.ubuntu.com/lunar/gcc)

https://packages.ubuntu.com/lunar/g++-13 has been installed as `sudo apt install g++13`.

fanquake commented:
With fa0d180f48d4e5253f58aced32768369f963d1c7 & `gcc (GCC) 13.1.1 20230426 (Red Hat 13.1.1-1)`, fixes all issues from #26926 except for:
```bash
test/interfaces_tests.cpp: In member function ‘void interfaces_tests::findCommonAncestor::test_method()’:
test/interfaces_tests.cpp:101:19: warning: possibly dangling reference to a temporary [-Wdangling-reference]
  101 |     const CChain& active = WITH_LOCK(Assert(m_node.chainman)->GetMutex(), return Assert(m_node.chainman)->ActiveChain());
      |                   ^~~~~~
In file included from ./common/args.h:9,
                 from ./test/util/setup_common.h:9,
                 from test/interfaces_tests.cpp:9:
./sync.h:302:96: note: the temporary was destroyed at the end of the full expression ‘<lambda closure object>interfaces_tests::findCommonAncestor::test_method()::<lambda()>{((interfaces_tests::findCommonAncestor*)this)}.interfaces_tests::findCommonAncestor::test_method()::<lambda()>()’
  302 | #define WITH_LOCK(cs, code) (MaybeCheckNotHeld(cs), [&]() -> decltype(auto) { LOCK(cs); code; }())
      |                                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
test/interfaces_tests.cpp:101:28: note: in expansion of macro ‘WITH_LOCK’
  101 |     const CChain& active = WITH_LOCK(Assert(m_node.chainman)->GetMutex(), return Assert(m_node.chainman)->ActiveChain());
      |                            ^~~~~~~~~
  CXX      test/test_bitcoin-netbase_tests.o
```

Commit:
Temporarily work around gcc-13 warning bug in interfaces_tests

This can be reverted once gcc excludes lambdas with decltype(auto)
return type from its -Wdangling-reference analysis.

stickies-v commented on src/test/interfaces_tests.cpp:
```diff
@@ -98,7 +98,7 @@ BOOST_AUTO_TEST_CASE(findAncestorByHash)
 {
     auto& chain = m_node.chain;
-    const CChain& active = WITH_LOCK(Assert(m_node.chainman)->GetMutex(), return Assert(m_node.chainman)->ActiveChain());
+    const CChain& active{*WITH_LOCK(Assert(m_node.chainman)->GetMutex(), return &Assert(m_node.chainman)->ActiveChain())};
```
I think it would be best to add the commit message as code documentation to make it easier to track when this can be reverted?

I'm also not sure this commit fits the scope of the PR but I can see how it's related, and it's a separate commit anyway so probably fine.

hebasto replied:
GCC 13 introduced a broken behavior, and we usually do not modify our correct code to make a broken compiler happy. Considering that this is the only line in test code, I lean to agree with this change.

hebasto replied:
As the test is execute in a single thread, is it big deal to hold `::cs_main` during the entire test case?

Why not to use the pattern as https://github.com/bitcoin/bitcoin/blob/fa266c4bbf564308ddbc12653527226506902084/src/test/interfaces_tests.cpp#L130-L132
?

hebasto replied:
> Likely due to a LOCKS_EXCLUDED?

Right, in `Chainstate::InvalidateBlock`.

hebasto reviewed (approved):
re-ACK fa266c4bbf564308ddbc12653527226506902084

hebasto commented:
Maybe notice the GCC-13 bug in the comments for `#define WITH_LOCK`?
